iml-services/iml-device/src/main.rs

```diff
@@ -86,31 +91,131 @@ async fn main() -> Result<(), ImlDeviceError> {
 
     tokio::spawn(server);
 
-    let mut s = consume_data::<Device>("rust_agent_device_rx");
+    let mut s = consume_data::<(Device, HashSet<Mount>)>("rust_agent_device_rx");
 
     let pool = iml_orm::pool().unwrap();
 
-    while let Some((f, d)) = s.try_next().await? {
+    while let Some((host, (devices, mounts))) = s.try_next().await? {
         let mut cache = cache2.lock().await;
 
-        cache.insert(f.clone(), d.clone());
+        cache.insert(host.clone(), devices.clone());
 
         let device_to_insert = NewChromaCoreDevice {
-            fqdn: f.to_string(),
-            devices: serde_json::to_value(d).expect("Could not convert incoming Devices to JSON."),
+            fqdn: host.to_string(),
+            devices: serde_json::to_value(devices)
+                .expect("Could not convert incoming Devices to JSON."),
         };
 
-        let new_device = diesel::insert_into(table)
+        let new_device = diesel::insert_into(dvc::table)
             .values(device_to_insert)
-            .on_conflict(fqdn)
+            .on_conflict(dvc::fqdn)
             .do_update()
-            .set(devices.eq(excluded(devices)))
+            .set(dvc::devices.eq(excluded(dvc::devices)))
             .get_result_async::<ChromaCoreDevice>(&pool)
             .await
             .expect("Error saving new device");
 
-        tracing::info!("Inserted device from host {}", new_device.fqdn);
-        tracing::trace!("Inserted device {:?}", new_device);
+        tracing::info!("Inserted device from host {}", host);
+        tracing::debug!("Inserted {:?}", new_device);
+
+        let lustre_mounts = mounts
+            .into_iter()
+            .filter(|m| m.fs_type.0.as_str() == "lustre")
+            .filter_map(|m| {
+                m.source
+                    .0
+                    .to_str()
+                    .and_then(|p| p.splitn(2, ":/").nth(1))
+                    .map(|fs| (fs.to_string(), m.target.0.to_str().map(String::from)))
+            })
+            .collect::<Vec<_>>();
+
+        tracing::debug!("Client mounts at {}: {:?}", host, &lustre_mounts);
+
+        let state_modified_at = chrono::offset::Utc::now().into_sql::<sql_types::Timestamptz>();
+        let not_deleted = Some(true).into_sql::<sql_types::Nullable<sql_types::Bool>>();
+        let cross = true.into_sql::<sql_types::Bool>();
+        let immutable_state = false.into_sql::<sql_types::Bool>();
+
+        let insert_mounts = lustre_mounts
+            .iter()
+            .map(|(fs, tg)| {
+                let mp = tg
+                    .clone()
+                    .into_sql::<sql_types::Nullable<sql_types::Text>>();
+                mgtfs::table
+                    .inner_join(mgthost::table.on(cross))
```

Comment:
This cross join is strange given there is no fk constraint between filesystems and hosts.

Reply:
That's the point of cross join. Actually since FS and host names are unique, I only get one record from each table.

Reply:
`chroma_core_lustreclientmount` has all the fk constraints in place already, which means there is no need for a cross join.

```
chroma=> \d chroma_core_lustreclientmount;
                                       Table "public.chroma_core_lustreclientmount"
      Column       |           Type           |                                 Modifiers
-------------------+--------------------------+----------------------------------------------------------------------------
 id                | integer                  | not null default nextval('chroma_core_lustreclientmount_id_seq'::regclass)
 state_modified_at | timestamp with time zone | not null
 state             | character varying(32)    | not null
 immutable_state   | boolean                  | not null
 not_deleted       | boolean                  |
 mountpoint        | character varying(1024)  |
 content_type_id   | integer                  |
 filesystem_id     | integer                  | not null
 host_id           | integer                  | not null
Indexes:
    "chroma_core_lustreclientmount_pkey" PRIMARY KEY, btree (id)
    "chroma_core_lustreclient_host_id_filesystem_id_no_b6406a2e_uniq" UNIQUE CONSTRAINT, btree (host_id, filesystem_id, not_deleted)
    "chroma_core_lustreclientmount_content_type_id_2ed6c185" btree (content_type_id)
    "chroma_core_lustreclientmount_filesystem_id_ce4463ef" btree (filesystem_id)
    "chroma_core_lustreclientmount_host_id_e5d16643" btree (host_id)
Foreign-key constraints:
    "chroma_core_lustrecl_content_type_id_2ed6c185_fk_django_co" FOREIGN KEY (content_type_id) REFERENCES django_content_type(id) DEFERRABLE INITIALLY DEFERRED
    "chroma_core_lustrecl_filesystem_id_ce4463ef_fk_chroma_co" FOREIGN KEY (filesystem_id) REFERENCES chroma_core_managedfilesystem(id) DEFERRABLE INITIALLY DEFERRED
    "chroma_core_lustrecl_host_id_e5d16643_fk_chroma_co" FOREIGN KEY (host_id) REFERENCES chroma_core_managedhost(id) DEFERRABLE INITIALLY DEFERRED
```

Reply:
Maybe it's better to show what I mean, technically, I select from a cross join:

```sql
INSERT INTO chroma_core_lustreclientmount
            (state_modified_at,
             state,
             immutable_state,
             not_deleted,
             mountpoint,
             content_type_id,
             filesystem_id,
             host_id)
(SELECT '2020-05-13 20:15:25.050399+00',
        'mounted',
        'f',
        't',
        '/tmp/fs2',
        35,
        fs.id,
        h.id
 FROM   chroma_core_managedfilesystem AS fs,
        chroma_core_managedhost AS h
 WHERE  fs.NAME = 'fs2'
        AND h.fqdn = 'c1.local')  
```